### PR TITLE
Allow UTF-8 characters in metric and label names as opt-in feature

### PR DESCRIPTION
**Pull request by ArthurSens**
Work originally comes from #14974

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -134,6 +134,10 @@ func LoadFile(filename string, agentMode bool, logger *slog.Logger) (*Config, er
 		}
 	}
 
+	if (model.NameValidationScheme != model.UTF8Validation) && cfg.OTLPConfig.TranslationStrategy == NoUTF8EscapingWithSuffixes {
+		return nil, errors.New("otlp translation strategy NoUTF8EscapingWithSuffixes is not allowed when UTF8 is disabled")
+	}
```
Comment (ArthurSens) [inline config/config.go:139]: I'm struggling a bit to find a good way to test this 🤔

**Comment by aknuds1** [comment]
Taking a look.

**Comment by aknuds1** [comment]
Could you fix conflicts?

**Review by aknuds1 — COMMENTED** [review]
Please see comments.

File: storage/remote/otlptranslator/prometheus/normalize_label_test.go [hunk 2]
```diff
@@ -23,22 +23,32 @@ import (
 func TestNormalizeLabel(t *testing.T) {
 	tests := []struct {
 		label    string
+		allowUTF bool
 		expected string
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply storage/remote/otlptranslator/prometheus/normalize_label_test.go:27]: good idea

File: config/config.go (hunk 1 above)
Comment (ArthurSens) [reply config/config.go]: got it

**Comment by ArthurSens** [comment]
Friendly ping @aknuds1, could we merge this now? :)

**Comment by aknuds1** [comment]
> Friendly ping @aknuds1, could we merge this now? :)

I'll have to review it first, getting around to it today.

File: config/config_test.go [hunk 3]
```diff
@@ -1553,6 +1554,18 @@ func TestOTLPSanitizeResourceAttributes(t *testing.T) {
 	})
 }
 
+func TestOTLPAllowUTF8(t *testing.T) {
+	t.Run("good config", func(t *testing.T) {
+		_, err := LoadFile(filepath.Join("testdata", "otlp_allow_utf8.good.yml"), false, promslog.NewNopLogger())
+		require.NoError(t, err)
```
Comment (aknuds1) [inline config/config_test.go:1560]: You should verify the outcome:
```suggestion
		c, err := LoadFile(filepath.Join("testdata", "otlp_allow_utf8.good.yml"), false, promslog.NewNopLogger())
		require.NoError(t, err)
		require.Equal(t, NoUTF8EscapingWithSuffixes, conf.OTLPConfig.TranslationStrategy)

```

File: storage/remote/write_handler.go [hunk 4]
```diff
@@ -513,6 +513,7 @@ func (h *otlpWriteHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	converter := otlptranslator.NewPrometheusConverter()
 	annots, err := converter.FromMetrics(r.Context(), req.Metrics(), otlptranslator.Settings{
 		AddMetricSuffixes:         true,
+		AllowUTF8:                 (model.NameValidationScheme == model.UTF8Validation) && (otlpCfg.TranslationStrategy == config.NoUTF8EscapingWithSuffixes),
```
Comment (aknuds1) [inline storage/remote/write_handler.go:516]: At this point, the name validation scheme should already be validated, and we shouldn't swallow a mistake in any case. If you want to make sure a programming error wasn't made, you should instead validate and return an error if the name validation scheme and translation strategy don't correspond to each other.
```suggestion
		AllowUTF8:                 otlpCfg.TranslationStrategy == config.NoUTF8EscapingWithSuffixes,
```

File: config/config.go [hunk 5]
```diff
@@ -134,6 +134,10 @@ func LoadFile(filename string, agentMode bool, logger *slog.Logger) (*Config, er
 		}
 	}
 
+	if cfg.GlobalConfig.MetricNameValidationScheme == "legacy" && cfg.OTLPConfig.TranslationStrategy == NoUTF8EscapingWithSuffixes {
+		return nil, errors.New("otlp translation strategy NoUTF8EscapingWithSuffixes is not allowed when UTF8 is disabled")
+	}
```
Comment (aknuds1) [inline config/config.go:139]: This check should be moved into `Load`, for completeness.

**Review by aknuds1 — CHANGES_REQUESTED** [review]
I found some more issues.

File: config/config.go [hunk 6]
```diff
@@ -134,6 +134,10 @@ func LoadFile(filename string, agentMode bool, logger *slog.Logger) (*Config, er
 		}
 	}
 
+	if cfg.GlobalConfig.MetricNameValidationScheme == "legacy" && cfg.OTLPConfig.TranslationStrategy == NoUTF8EscapingWithSuffixes {
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply config/config.go:137]: good catch

File: config/config.go (hunk 5 above)
Comment (aknuds1) [reply config/config.go]: It's not strange @ArthurSens. The other validation in `LoadFile` is dependent on `agentMode`, which is a parameter only to `LoadFile` and not `Load`.

Comment (ArthurSens) [reply config/config.go]: moved...

File: documentation/examples/prometheus-otlp.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+otlp:
+  # Recommended attributes to be promoted to labels.
+  promote_resource_attributes:
+    - service.instance.id
+    - service.name
+    - service.namespace
+    - cloud.availability_zone
+    - cloud.region
+    - container.name
+    - deployment.environment
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply documentation/examples/prometheus-otlp.yml:14]: I made that fix in my branch, that I've proposed to Arthur, @jesusvazquez

File: storage/remote/otlptranslator/prometheus/normalize_name.go [hunk 8]
```diff
@@ -88,45 +88,51 @@ var perUnitMap = map[string]string{
...
+	}
 
 	// Namespace?
 	if namespace != "" {
 		return namespace + "_" + metricName
 	}
 
-	// Metric name starts with a digit? Prefix it with an underscore.
-	if metricName != "" && unicode.IsDigit(rune(metricName[0])) {
+	// Metric name starts with a digit and utf8 not allowed? Prefix it with an underscore.
```
Comment (aknuds1) [inline storage/remote/otlptranslator/prometheus/normalize_name.go:115]: @jesusvazquez do you know if this change is correct?

File: storage/remote/otlptranslator/prometheus/normalize_name_test.go [hunk 9]
```diff
@@ -25,92 +25,101 @@ import (
...
 }
 
-func TestUnsupportedRunes(t *testing.T) {
-	require.Equal(t, "unsupported_metric_temperature_F", normalizeName(createGauge("unsupported.metric.temperature", "°F"), ""))
-	require.Equal(t, "unsupported_metric_weird", normalizeName(createGauge("unsupported.metric.weird", "+=.:,!* & #"), ""))
-	require.Equal(t, "unsupported_metric_redundant_test_per_C", normalizeName(createGauge("unsupported.metric.redundant", "__test $/°C"), ""))
+func TestAllowUTF8(t *testing.T) {
+	t.Run("allow UTF8", func(t *testing.T) {
+		require.Equal(t, "unsupported.metric.temperature_°F", normalizeName(createGauge("unsupported.metric.temperature", "°F"), "", true))
+		require.Equal(t, "unsupported.metric.weird_+=.:,!* & #", normalizeName(createGauge("unsupported.metric.weird", "+=.:,!* & #"), "", true))
```
Comment (aknuds1) [inline storage/remote/otlptranslator/prometheus/normalize_name_test.go:88]: If you try applying the following, it's going to fail. There's a fundamental bug in the UTF-8 handling.
```suggestion
		require.Equal(t, "unsupported.metric.+weird_+=.:,!* & #", normalizeName(createGauge("unsupported.metric.+weird", "+=.:,!* & #"), "", true))
```

File: storage/remote/otlptranslator/prometheus/normalize_name.go (hunk 8 above)
Comment (jesusvazquez) [reply storage/remote/otlptranslator/prometheus/normalize_name.go:115]: I have no clue 😱

File: storage/remote/otlptranslator/prometheus/normalize_name_test.go (hunk 9 above)
Comment (ArthurSens) [reply storage/remote/otlptranslator/prometheus/normalize_name_test.go:88]: Yes you're correct.

Just wanted to mention that although OTel allows full UTF8, the most common ones are dots and foreign characters, and this PR is focused to unblock OTel

Comment (aknuds1) [reply storage/remote/otlptranslator/prometheus/normalize_name_test.go:88]: I don't think we should approve a PR with fundamentally broken logic.

File: storage/remote/otlptranslator/prometheus/normalize_name_test.go [hunk 10]
```diff
@@ -25,92 +25,119 @@ import (
...
+	//
+	// Now that we're allowing OTel users to have their original names store in prometheus without any transformation, we're facing problems
+	// where two (or more) UTF-8 characters are being used to separate words.
+	// TODO(arthursens): Fix it!
+
+	// We're asserting on 'NotEqual', which proves the bug.
+	require.NotEqual(t, "metric....split_=+by_//utf8characters", normalizeName(createGauge("metric....split_=+by_//utf8characters", ""), "", true))
+	// Here we're asserting on 'Equal', showing the current behavior
+	require.Equal(t, "metric.split_by_utf8characters", normalizeName(createGauge("metric....split_=+by_//utf8characters", ""), "", true))
 }
```
Comment (ArthurSens) [inline storage/remote/otlptranslator/prometheus/normalize_name_test.go:116]: I've added a test showcasing the known bug, along with the explanation to why we've got to this situation.

This is a super edge case since appliations using the OTel SDK follow [semantic convention](https://opentelemetry.io/docs/concepts/semantic-conventions/), and those conventions do not have metrics with repeated dots or things like that. 

Could we tackle this bug in a separate PR since it's not mission critical?

**Review by jesusvazquez — APPROVED** [review]
LGTM  lets do a close followup on that bug in case it happens in unknown cases!

**Comment by bwplotka** [comment]
FYI, porting to 3.0 https://github.com/prometheus/prometheus/pull/15384
